### PR TITLE
prepare the 3 important db packages for ts

### PR DESCRIPTION
**Pull request by foxriver76**
ported db-objects-redis/objectsUtils to ts, because it had compile errors due to the strange WMStrm construct, please have a look at this file during review. Mostly I ve used any for now, because with real porting db itself many typing will be inferred then

### CODE REVIEW

File: packages/db-objects-redis/package.json [hunk 1]
```diff
@@ -26,14 +26,19 @@
     "type": "git",
     "url": "https://github.com/ioBroker/ioBroker.js-controller/packages/db-objects-redis"
   },
-  "scripts": {},
-  "main": "index.js",
+  "scripts": {
+    "build": "tsc -b tsconfig.build.json && npm run copy-files",
+    "copy-files": "cp -r ./src/lib/objects/lua* ./build/lib/objects/"
```
Comment (AlCalzone) [inline packages/db-objects-redis/package.json:31]:
```suggestion
    "build": "tsc -b tsconfig.build.json",
    "postbuild": "cp -r ./src/lib/objects/lua* ./build/lib/objects/"
```
seems a bit simpler :)

File: packages/db-objects-redis/src/lib/objects/constants.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+export const ACCESS_GROUP_WRITE = 0x20;
+export const ACCESS_GROUP_READ = 0x40;
+export const ACCESS_GROUP_RW = ACCESS_GROUP_WRITE | ACCESS_GROUP_READ;
+export const ACCESS_GROUP_ALL = ACCESS_GROUP_WRITE | ACCESS_GROUP_READ | ACCESS_GROUP_EXEC;
+
+export const ACCESS_USER_EXEC = 0x100;
+export const ACCESS_USER_WRITE = 0x200;
+export const ACCESS_USER_READ = 0x400;
+export const ACCESS_USER_RW = ACCESS_USER_WRITE | ACCESS_USER_READ;
+export const ACCESS_USER_ALL = ACCESS_USER_WRITE | ACCESS_USER_READ | ACCESS_USER_EXEC;
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/constants.ts:26]: We could think about turning these into `enum`s

**Review by AlCalzone — COMMENTED** [review]
So far so good, just some minor things

File: packages/db-objects-redis/src/lib/objects/constants.ts (hunk 2 above)
Comment (foxriver76) [reply packages/db-objects-redis/src/lib/objects/constants.ts:26]: Problem is that or logic isn’t working in enums so would have to work around but yes can change 👍

File: packages/db-objects-redis/package.json (hunk 1 above)
Comment (foxriver76) [reply packages/db-objects-redis/package.json:31]: Ah Nice will change it in adapter too where we copy certs

File: packages/db-objects-redis/src/lib/objects/constants.ts (hunk 2 above)
Comment (AlCalzone) [reply packages/db-objects-redis/src/lib/objects/constants.ts:26]: Yeah, this is not a priority - just an idea for a future cleanup. The usage sites would have to be updated aswell of course.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts [hunk 3]
```diff
@@ -574,12 +579,18 @@ export function sanitizePath(id: string, name: string) {
...
+    options: Record<string, any>,
+    flag: CONSTS.GenericAccessFlags
+): boolean {
     // read rights of object
     if (!obj || !obj.common || !obj.acl || flag === CONSTS.ACCESS_LIST) {
         return true;
     }
 
+    // FIXME: what if flag is ACCESS_DELETE or ACCESS_CREATE?
+
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsUtils.ts:593]: We currently don't handle this case, but its theoretically possible to call the method this way (I think we already do!)

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts [hunk 4]
```diff
@@ -329,41 +333,42 @@ export function checkFileRights(
...
+            create: true,
+            list: true
+        }
     };
-
-    return acl;
 }
 
-export function getUserGroup(objects: any, user: string, callback: () => void) {
+// FIXME: Not sure about the expected return type here? ioBroker.GroupObject maybe? Or an array thereof?
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsUtils.ts:370]: This should have a clear return type, but I currently can't decipher it.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts [hunk 5]
```diff
@@ -329,41 +333,42 @@ export function checkFileRights(
...
-        write: true,
-        delete: true
-    };
-    acl.state = {
-        read: true,
-        write: true,
-        delete: true,
-        create: true,
-        list: true
+function getDefaultAdminRights(acl?: ioBroker.ObjectPermissions, _isState?: boolean) {
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsUtils.ts:336]: All the existing structures for permissions/ACL/etc. are pretty messy. I did capture most of them here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/20f9baa9f3dbf1eb5cfc5c3c7d974b92b211fc61/types/iobroker/index.d.ts#L68-L104
but I believe there are other variants in use too.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts [hunk 6]
```diff
@@ -615,14 +626,15 @@ export function checkObject(obj: ioBroker.Object, options: Record<string, any>,
     return true; // ALL OK
 }
 
+// FIXME: Return type?
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsUtils.ts:629]: It feels like this should have an easy return type, but the control flow isn't clean. There's a mix of `return tools.maybeCallback...` on the root level and inside callbacks.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts [hunk 7]
```diff
@@ -244,9 +248,9 @@ export function checkFileRights(
     id: string,
     name: string,
     options: Record<string, any>,
-    flag: any,
+    flag: CONSTS.GenericAccessFlags,
     callback: () => void
-) {
+): any {
```
Comment (AlCalzone) [inline packages/db-objects-redis/src/lib/objects/objectsUtils.ts:253]: Also not sure about the return type here.

**Review by AlCalzone — CHANGES_REQUESTED** [review]

**Comment by AlCalzone** [comment]
I cleaned up most of the `any`s, but I'm not sure about all of them. And there are a few things that are currently not correct and need to be fixed IMO.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts (hunk 5 above)
Comment (AlCalzone) [reply packages/db-objects-redis/src/lib/objects/objectsUtils.ts:336]: Not sure if `ObjectPermissions` is actually the intended type here.

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts (hunk 4 above)
Comment (foxriver76) [reply packages/db-objects-redis/src/lib/objects/objectsUtils.ts]: I've added it from my understanding

File: packages/db-objects-redis/src/lib/objects/objectsUtils.ts (hunk 6 above)
Comment (foxriver76) [reply packages/db-objects-redis/src/lib/objects/objectsUtils.ts]: To achieve 100 % clean handling we would need to make getUserGroup in db itself async or in utils and map to utils, but db is exposed so not sure if we should do it right now

**Comment by foxriver76** [comment]
do we want to incorporate as it is now and fix other types later if db-base is ported? @AlCalzone @Apollon77

**Comment by AlCalzone** [comment]
Fine with me.

**Pull request merged**